tiledb/common/random/random_label.h

```diff
@@ -33,10 +33,90 @@
 #ifndef TILEDB_HELPERS_H
 #define TILEDB_HELPERS_H
 
+#include <iomanip>
+#include <mutex>
+#include <sstream>
 #include <string>
 
+#include "tiledb/common/exception/exception.h"
+#include "tiledb/common/random/prng.h"
+#include "tiledb/sm/misc/tdb_time.h"
+
 namespace tiledb::common {
 
+class RandomLabelException : public StatusException {
+ public:
+  explicit RandomLabelException(const std::string& message)
+      : StatusException("RandomLabel", message) {
+  }
+};
+
+class RandomLabelGenerator {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+  DISABLE_COPY_AND_COPY_ASSIGN(RandomLabelGenerator);
+  DISABLE_MOVE_AND_MOVE_ASSIGN(RandomLabelGenerator);
+
+  /** Default destructor. */
+  ~RandomLabelGenerator() = default;
+
+ protected:
+  /** Protected constructor, abstracted by public-facing accessor. */
+  RandomLabelGenerator();
+
+  /* ********************************* */
+  /*                API                */
+  /* ********************************* */
+  /** Generate a random label. */
+  std::string generate() {
+    PRNG& prng = PRNG::get();
+    std::lock_guard<std::mutex> lock(mtx_);
+    auto now = tiledb::sm::utils::time::timestamp_now_ms();
+
+    // If no label has been generated this millisecond, generate a new one.
+    if (now != prev_time_) {
+      prev_time_ = now;
+      counter_ = static_cast<uint32_t>(prng());
+      // Clear the top bit of the counter such that a full 2 billion values
+      // could be generated within a single millisecond.
+      counter_ &= 0x7FFFFFFF;
+    } else {
+      counter_ += 1;
+      if (counter_ == 0) {
+        throw RandomLabelException("Maximum generation frequency exceeded.");
+      }
```

Comment:
Sleep for one millisecond and retry?

Reply:
Nah, if we ever manage to generate 2 billion UUIDs in a second, we'll want to know about it rather than paper over that error.

Reply:
Oh yeah we clear the top bit of the counter so we can't overflow by chance.